ARROW-8287: [Rust] Add "pretty" util to help with printing tabular output of RecordBatches

markhildreth opened the pull request:
Just a simple move of code from DataFusion to Arrow, and using it. I have a few comments/questions on this [which I have put on the JIRA issue](https://issues.apache.org/jira/browse/ARROW-8287?focusedCommentId=17086534&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17086534).

andygrove commented:
Thanks @markhildreth for the detailed write-up in the JIRA! I've started looking through this. 

I'm not sure I understand all the points you made yet, but if there is a way to just expose the print_batches method and keep the pretty table details internal to utils then we should do that.

I also think it is fine for the moment to ignore the use case where the schema varies between record batches and file a separate issue for that.

nevi-me commented:
> I also think it is fine for the moment to ignore the use case where the schema varies between record batches and file a separate issue for that.

Just on this point, there shouldn't be a way of creating a record batch that has a mismatching schema

markhildreth commented:
@andygrove Thanks for the feedback. I have updated the PR with a less leaky API. I also fixed the type inference problem that was caused by the new dependency.

@nevi-me True for a singular batch, but it's possible to create multiple batches from different sources. For example

```
let schema1 = ...
let csv1 = ...
let batch1 = ...

let schema2 = ... // different schema
let csv2 = ...
let batch2 = ...

print_batches(&[batch1, batch2]);
```

As I said, this is probably not something to worry about too much right now, but I'll probably add an issue for it for later if that's alright. Interestingly, this code wouldn't even necessarily crash; you would just get an odd-looking table:

```
+---------------------------------------+-----------+-----------+-------+
| city                                  | lat       | lng       |       |
+---------------------------------------+-----------+-----------+-------+
| Elgin, Scotland, the UK               | 57.653484 | -3.335724 |       |
| Stoke-on-Trent, Staffordshire, the UK | 53.002666 | -2.179404 |       |
| Solihull, Birmingham, UK              | 52.412811 | -1.778197 |       |
| Cardiff, Cardiff county, UK           | 51.481583 | -3.17909  |       |
| 1                                     | 1.1       | 1.11      | true  |
| 2                                     | 2.2       | 2.22      | true  |
| 3                                     | 0         | 3.33      | true  |
| 4                                     | 4.4       |           | false |
| 5                                     | 6.6       |           | false |
+---------------------------------------+-----------+-----------+-------+

```

markhildreth commented on rust/parquet/src/encodings/rle.rs:
```diff
@@ -830,7 +826,7 @@ mod tests {
             let mut rng = thread_rng();
             let seed_vec: Vec<u8> =
-                Standard.sample_iter(&mut rng).take(seed_len).collect();
+                rng.sample_iter::<u8, _>(&Standard).take(seed_len).collect();
```
This is an example of type inferences that would become compiler errors with the inclusion of the `prettytable-rs` crate. I have fixed this code, but the original error being generated was:

```
error[E0282]: type annotations needed
   --> parquet/src/encodings/rle.rs:833:26
    |
833 |                 Standard.sample_iter(&mut rng).take(seed_len).collect();
    |                          ^^^^^^^^^^^ cannot infer type for `T`
```

markhildreth commented:
From a purely practical standpoint, this PR is ready for further review and merging. If approved, I would probably add some minor issue for the following:
* Trying to avoid the type inference issue.
* Creating a type that can be used to iterate over multiple batches with statically guaranteed same schemas.

Additionally, I’ll create a new issue and PR to cover moving DataFusion to this using this API.

andygrove commented:
@markhildreth This looks great, but is now duplicating the code between arrow and datafusion. Can we remove the datafusion utils copy and have datafusion use the arrow utils instead?

markhildreth commented:
@andygrove Yup, I was planning on doing that in a separate PR. If you'd like I can do that in this one.

markhildreth commented:
Created [follow-up JIRA task](https://issues.apache.org/jira/browse/ARROW-8590).
